8276184: Exclude lambda proxy class from the CDS archive if its caller class is excluded

calvinccheung opened the pull request:
The proposed patch has been redone completely to address the problem when a caller class is being excluded
from the archive, the associated lambda proxy class(es) was not excluded.

The caller class could be excluded if:
- it is signed;
- it has been redefined after its lambda proxy class was created.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276184](https://bugs.openjdk.java.net/browse/JDK-8276184): Exclude lambda proxy class from the CDS archive if its caller class is excluded


### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6205/head:pull/6205` \
`$ git checkout pull/6205`

Update a local copy of the PR: \
`$ git checkout pull/6205` \
`$ git pull https://git.openjdk.java.net/jdk pull/6205/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6205`

View PR using the GUI difftool: \
`$ git pr show -t 6205`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6205.diff">https://git.openjdk.java.net/jdk/pull/6205.diff</a>

</details>

calvinccheung commented:
/label add hotspot-runtime

iklam commented:
Hi Calvin, the reason I suggested adding more tests for lambda proxies (for JDK-8274944) is for these two cases in SystemDictionaryShared::check_for_exclusion_impl():

```
  if (has_been_redefined(k)) {
    return warn_excluded(k, "Has been redefined");    <<<<<< HERE
  }
  if (k->signers() != NULL) {
    // We cannot include signed classes in the archive because the certificates
    // used during dump time may be different than those used during
    // runtime (due to expiration, etc).
    return warn_excluded(k, "Signed JAR");  <<<< HERE
  }
```

The second case should be pretty easy to test for:

```
class SignedFoo {
    public static void main(final String... args) {
        System.out.println(getRunnable());
    }

    public static Runnable getRunnable() {
        return () -> {};
    }
}

$ java -cp . SignedFoo
SignedFoo$$Lambda$1/0x0000000800c009f0@7344699f
```
Since we don't archive SignedFoo, we should also not archive the lambda proxy class generated for SignedFoo.

The first case would be a bit more involved:

```
class RedefinedFoo {
    public static void main(final String... args) {
        System.out.println(getRunnable());   // line A
        try_to_redefine_this_class();              // line B
    }

    public static Runnable getRunnable() {
        return () -> {};
    }
}
```

At line A, the lambda proxy is generated. At line B, InstanceKlass::has_been_redefined() will return true for RedefinedFoo. This should cause RedefinedFoo to be excluded from the archive, and thus the lambda proxy should be excluded as well.

For these test cases, I think it would be better to use dynamic dump, since it's legal for program to perform these actions while running with -XX:ArchiveClassesAtExit=foo.jsa.

iklam commented:
FYI, I applied this patch and ran test/hotspot/jtreg/runtime/cds/appcds/SignedJar.java:

```
$ git diff
diff --git a/test/hotspot/jtreg/runtime/cds/appcds/test-classes/Hello.java b/test/hotspot/jtreg/runtime/cds/appcds/test-classes/Hello.java
index 250a63adde8..f421ba5477f 100644
--- a/test/hotspot/jtreg/runtime/cds/appcds/test-classes/Hello.java
+++ b/test/hotspot/jtreg/runtime/cds/appcds/test-classes/Hello.java
@@ -25,5 +25,10 @@
 public class Hello {
   public static void main(String args[]) {
     System.out.println("Hello World");
+    System.out.println(getRunnable());
+  }
+
+  public static Runnable getRunnable() {
+    return () -> {};
   }
 }
```

Inside the JTREG scratch directory, you will find hello.jar and signed_hello.jar. The output of the second command seems suspicious. The Hello class is not archived, but its lambda proxy class is archived:

```
$ java -cp hello.jar -XX:ArchiveClassesAtExit=foo.jsa -Xlog:cds+class=debug Hello | grep Hello
Hello World
Hello$$Lambda$1/0x0000000800c009f0@18b4aac2
[0.144s][debug][cds,class] klasses[    1] = 0x0000000800bb4638 app   Hello$$Lambda$1/0x0000000800c009f0 ** hidden
[0.144s][debug][cds,class] klasses[    2] = 0x0000000800bb49e0 app   Hello

$ java -cp signed_hello.jar -XX:ArchiveClassesAtExit=foo.jsa -Xlog:cds+class=debug Hello | grep Hello
Hello World
Hello$$Lambda$11/0x0000000800c011f0@5eb5c224
[0.257s][warning][cds] Skipping Hello: Signed JAR
[0.289s][debug  ][cds,class] klasses[  103] = 0x0000000800bea278 app   Hello$$Lambda$11/0x0000000800c011f0 ** hidden
```

calvinccheung commented:
Hi Ioi,
Thanks for your review and noticing the bug.
I've reworked the patch and added the test cases as you suggested.
The bug and PR title and description have been updated to reflect the bug that is being addressed.
thanks,
Calvin

iklam commented on src/hotspot/share/classfile/systemDictionaryShared.hpp:
```diff
@@ -188,6 +187,8 @@ class SystemDictionaryShared: public SystemDictionary {
                        outputStream* st) NOT_CDS_RETURN;
 
 public:
+  static bool check_for_exclusion_impl(InstanceKlass* k, bool silent);
```
Since check_for_exclusion_impl is used only by CleanupDumpTimeLambdaProxyClassTable, I think it's better to keep it private, but declare CleanupDumpTimeLambdaProxyClassTable as a friend class of SystemDictionaryShared. We already do that for ExcludeDumpTimeSharedClasses.

calvinccheung replied:
Fixed.

iklam commented on src/hotspot/share/classfile/systemDictionaryShared.cpp:
```diff
@@ -1602,9 +1612,19 @@ class CleanupDumpTimeLambdaProxyClassTable: StackObj {
     assert_lock_strong(DumpTimeTable_lock);
+    InstanceKlass* caller_ik = key.caller_ik();
+    if (SystemDictionaryShared::check_for_exclusion_impl(caller_ik, true)) {
+      // If the caller class is excluded, unregister all the associated lambda proxy classes so that
```
As we discussed offline, I think we should add a comment here that SystemDictionaryShared::is_excluded_class() cannot be called because with dynamic dumps, caller_ik could be a class in the base archive.

Also, I think we should test different combinations of how a proxy class can be excluded by the two conditions in this function:

(1) caller_ik is excluded
(2) proxy itself is excluded (because it implements an interface with an old class version
(3) both (1) and (2) are true

calvinccheung replied:
Added a comment.
The RedefineCallerClassTest.java has been expanded to cover the above 3 cases.

iklam commented on test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/RedefineCallerClassTest.java:
```diff
@@ -0,0 +1,83 @@
+ * @test
+ * @summary If the caller class is redefined during dump time, the caller class
+ *          and its lambda proxy class should not be archived.
+ * @requires vm.cds
```
Add `@bug .....`

calvinccheung replied:
Fixed.

iklam commented on src/hotspot/share/classfile/systemDictionaryShared.cpp:
```diff
@@ -1602,9 +1612,19 @@ class CleanupDumpTimeLambdaProxyClassTable: StackObj {
     for (int i = 0; i < info._proxy_klasses->length(); i++) {
       InstanceKlass* ik = info._proxy_klasses->at(i);
-      if (!ik->can_be_verified_at_dumptime()) {
+      if (SystemDictionaryShared::check_for_exclusion_impl(ik, true)) {
```
With the original code, if the list has two elements, only the first one is removed -- after the first iteration, `i` becomes 1, but `_proxy_klasses->length()` is also 1, and thus the loop will terminate.

I think the loop should be reversed:
```
for (int i = info._proxy_klasses->length() - 1; i >= 0; i--) {
  if (...) {
    info._proxy_klasses->remove_at(i);
  }
}
```

For test coverage, we should also have two test case:

- caller class Foo1 that uses a single instance of an lambda based on old interface
- caller class Foo2 that uses two instances of an lambda based on old interface

calvinccheung replied:
Nice catch! I've fixed the code.
I've expanded the existing LambdaContainsOldInf.java test to cover the above cases.

iklam reviewed (commented):
Looks good. I have some comments about a potential bug and suggestion for more testing.

Also, SystemDictionaryShared::cleanup_lambda_proxy_class_dictionary() is called separately for static and dynamic dumps. I think we should move this function to the end of 
SystemDictionaryShared::check_excluded_classes().

calvinccheung commented:
> Looks good. I have some comments about a potential bug and suggestion for more testing.
> 
> Also, SystemDictionaryShared::cleanup_lambda_proxy_class_dictionary() is called separately for static and dynamic dumps. I think we should move this function to the end of SystemDictionaryShared::check_excluded_classes().

Fixed. Also changed the declaration of `SystemDictionaryShared::cleanup_lambda_proxy_class_dictionary()` to private.

calvinccheung commented on test/hotspot/jtreg/runtime/cds/appcds/test-classes/Hello.java:
```diff
@@ -25,5 +25,10 @@
 public class Hello {
   public static void main(String args[]) {
     System.out.println("Hello World");
+    System.out.println(getRunnable());
```
Fixed.

iklam reviewed (approved):
LGTM. Just a small suggestion for an additional message check.

calvinccheung commented on test/hotspot/jtreg/runtime/cds/appcds/SignedJar.java:
```diff
@@ -42,24 +43,31 @@ public static void main(String[] args) throws Exception {
         // Test class exists in both signed JAR and unsigned JAR
         String jars = signedJar + System.getProperty("path.separator") + unsignedJar;
-        output = TestCommon.dump(jars, TestCommon.list("Hello"));
+        output = TestCommon.dump(jars, TestCommon.list(mainClass));
```
Not in this case. When the `do_entry()` of `ExcludeDumpTimeSharedClasses` is called, it doesn't know that the lambda proxy is associated with a signed `caller_ik`. If there's a need for a log message, it needs to be added in the `do_entry()` of `CleanupDumpTimeLambdaProxyClassTable`.

iklam replied:
Checking the exclusion message, even if it's printed, will be more involved, especially if you want to ensure that several such proxy classes are excluded. Since the proxies have similar but non-deterministic names (they have numbers in them), you will need to search for a pattern in the log and count the number of times it appears.

I think it's easier and more reliable to run with `-Xlog:cds+class=debug`, and check that the proxy class is not in the print out.

calvinccheung replied:
There's a bug in Hello.java which I've fixed in my latest push.
Also updated SignedJar.java to perform more checks on the output.

iklam reviewed (commented):
New test looks good.

dholmes-ora reviewed (commented):
Hi Calvin,

I really don't like the existing style of this code in relation to warn_excluded and adding the additional "silent" parameter just makes this worse in my opinion. It is not clear why you only want to log a warning in some cases.

David

calvinccheung commented:
> Hi Calvin,
> 
> I really don't like the existing style of this code in relation to warn_excluded and adding the additional "silent" parameter just makes this worse in my opinion. It is not clear why you only want to log a warning in some cases.
> 
> David

Hi David,

Thanks for taking a look.
It is to avoid logging the warning messages more than once.

thanks,
Calvin

calvinccheung commented on src/hotspot/share/classfile/systemDictionaryShared.cpp:
```diff
@@ -214,7 +214,7 @@ bool SystemDictionaryShared::check_for_exclusion(InstanceKlass* k, DumpTimeClass
 
   if (!info->has_checked_exclusion()) {
-    if (check_for_exclusion_impl(k)) {
+    if (check_for_exclusion_impl(k, false)) {
```
I will add a comment.

calvinccheung commented on src/hotspot/share/classfile/systemDictionaryShared.cpp:
```diff
@@ -224,9 +224,11 @@ bool SystemDictionaryShared::check_for_exclusion(InstanceKlass* k, DumpTimeClass
-bool SystemDictionaryShared::warn_excluded(InstanceKlass* k, const char* reason) {
-  ResourceMark rm;
-  log_warning(cds)("Skipping %s: %s", k->name()->as_C_string(), reason);
+bool SystemDictionaryShared::warn_excluded(InstanceKlass* k, const char* reason, bool silent) {
```
What if I leave the `warn_excluded()` function as before and make some change at the call sites. See below.

calvinccheung commented on src/hotspot/share/classfile/systemDictionaryShared.cpp:
```diff
@@ -259,41 +269,41 @@ bool SystemDictionaryShared::is_hidden_lambda_proxy(InstanceKlass* ik) {
+bool SystemDictionaryShared::check_for_exclusion_impl(InstanceKlass* k, bool silent) {
   if (k->is_in_error_state()) {
-    return warn_excluded(k, "In error state");
+    return warn_excluded(k, "In error state", silent);
```
The above return statement would become:
    `return silent ? true : warn_excluded(k, "In error state");`
What do you think?

dholmes-ora commented:
Why would you exclude a class more than once?

It still seems a very awkward way to code this.

And the existing use of `warn_excluded` as a proxy for `return true` is not good.

calvinccheung commented:
> Why would you exclude a class more than once?
> 
Consider a caller_ik which is signed. The first time it was encountered for exclusion is in the `do_entry()` of `ExcludeDumpTimeSharedClasses`. The second time it was encountered is in `SystemDictionaryShared::cleanup_lambda_proxy_class_dictionary()`.
> It still seems a very awkward way to code this.
> 
> And the existing use of `warn_excluded` as a proxy for `return true` is not good.

If I change `warn_excluded` to a void function, the call sites in `check_for_exclusion_impl()` would be like the following:
```
if (!silent) {
    warn_excluded(...);
}
return true;
```

calvinccheung commented:
@iklam I've pushed another update with the changes you suggested.
As we've discussed offline, the change also handles the case when an old interface/class is already in the base archive,
the interface/class should be considered verified during dump time. Therefore, we should also archive a class which
references the old interface/class in the base archive during dynamic dump.
Added 2 test cases for the above.

iklam reviewed (approved):
The latest version looks good to me.

dholmes-ora reviewed (approved):
Hi Calvin,

This versions seems much better - thanks.

I still really dislike `SystemDictionaryShared::warn_excluded` always returning true but that's not an issue for this PR. :)

Thanks,
David

calvinccheung commented:
Thanks @dholmes-ora and @iklam for the review.

calvinccheung commented:
/integrate

iklam commented:
> Hi Calvin,
> 
> This versions seems much better - thanks.
> 
> I still really dislike `SystemDictionaryShared::warn_excluded` always returning true but that's not an issue for this PR. :)
> 
> Thanks, David

It's a matter of style. The current syntax is more concise, and avoids coding errors like these:

```
    if (...) {
       warn_excluded("some message");
       // forgets to return
    }
    if (...) {
       warn_excluded("some message");
       return false; // returns the wrong value
    }
```
